fms_extras/utils/generation.py

```diff
@@ -5,7 +5,282 @@
 import torch
 import torch.nn.functional as F
 
-from fms_extras.utils.cache.paged import PagedKVCacheManager
+from fms_extras.models.speculator import (
+    MLPSpeculator,
+    flatten_batch,
+    select_inflate_dim,
+)
+from fms_extras.utils.cache.paged import PagedAttentionCacheData, PagedKVCacheManager
+
+
+def speculative_generate(
```

Comment:
I think we should try to refactor this into some helper functions since it's hundreds of LOC in one function. it's harder to follow the control flow of the nested for loops when the body of the loops doesn't fit in a window (need to page-down about 4 times to see the full body of the outer loop)

Reply:
we can probably also add a unit test similar to:

https://github.com/foundation-model-stack/foundation-model-stack/blob/main/tests/utils/test_generate.py